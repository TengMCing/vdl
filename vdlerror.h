//
// Created by Patrick Li on 9/11/22.
//

#ifndef VDL_VDLERROR_H
#define VDL_VDLERROR_H

#include <stdio.h>
#include <stdlib.h>

/*-----------------------------------------------------------------------------
 |  Error code
 ----------------------------------------------------------------------------*/

/// @description Error code.
typedef enum VDL_ERROR_CODE
{
    VDL_ERROR_RESOLVED                 = 0,
    VDL_ERROR_EXCEED_STACK_DEPTH_LIMIT = 1,
    VDL_ERROR_NULL_POINTER             = 2,
    VDL_ERROR_INDEX_OUT_OF_BOUND       = 3,
    VDL_ERROR_UNKNOWN_TYPE             = 4,
    VDL_ERROR_NOT_STACK_ALLOCATED      = 5,
    VDL_ERROR_NOT_HEAP_ALLOCATED       = 6,
    VDL_ERROR_INCOMPATIBLE_TYPE        = 7,
    VDL_ERROR_ZERO_CAPACITY            = 8,
    VDL_ERROR_ZERO_LENGTH_VECTOR       = 9,
    VDL_ERROR_ZERO_OBJECTS             = 10
} VDL_ERROR_CODE;

/*-----------------------------------------------------------------------------
 |  Error definition
 ----------------------------------------------------------------------------*/

/// @description Global error code.
static VDL_ERROR_CODE VDL_INTERNAL_GERROR = VDL_ERROR_RESOLVED;

/// @description Get a copy of the global error code.
#define vdl_GetError() ((VDL_ERROR_CODE){VDL_INTERNAL_GERROR})

/*-----------------------------------------------------------------------------
 |  Throw an error
 ----------------------------------------------------------------------------*/

#define VDL_THROW goto VDL_EXCEPTION

/*-----------------------------------------------------------------------------
 |  Check error
 ----------------------------------------------------------------------------*/

/// @description Check error code and jump to VDL_EXCEPTION if needed.
#define vdl_internal_CheckError()                      \
    do {                                               \
        if (VDL_INTERNAL_GERROR != VDL_ERROR_RESOLVED) \
            VDL_THROW;                                 \
    } while (0)

/*-----------------------------------------------------------------------------
 |  Except
 ----------------------------------------------------------------------------*/

/// @description Exception handler. Use it as a context manager.
/// @param error_code (int). The target error code.
#define vdl_Except(error_code) for (int _except = VDL_INTERNAL_GERROR != (error_code); _except < 1; VDL_INTERNAL_GERROR = VDL_ERROR_RESOLVED, _except++)

/// @description Final exception handler. Use it as a context manager.
#define vdl_Finally() for (int _except = 0; _except < 1; VDL_INTERNAL_GERROR = VDL_ERROR_RESOLVED, _except++)

/*-----------------------------------------------------------------------------
 |  Exit
 ----------------------------------------------------------------------------*/

/// @description Abort the program.
_Noreturn static inline void vdl_Abort(void)
{
    printf("Program aborted!\n");
    exit(EXIT_FAILURE);
}

#endif//VDL_VDLERROR_H
