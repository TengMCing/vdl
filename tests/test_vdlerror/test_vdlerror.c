//
// Created by Patrick Li on 9/11/22.
//

#include "../../vdlerror.h"

#include "../test.h"

int test_CheckError(void);
void test_Abort(void);

int main(void)
{
    // echo
    echo("Test vdl_GetError:");
    // expect(True)
    if (vdl_GetError() == VDL_INTERNAL_GERROR)
        test_printf("True");
    else
        test_printf("False");

    // echo
    echo("Test vdl_internal_CheckError, vdl_Except and vdl_Finally:");
    test_CheckError();

    test_Abort();
    return 0;
}

int test_CheckError(void)
{
    VDL_INTERNAL_GERROR = 1;
    vdl_internal_CheckError();
VDL_EXCEPTION:
    // expect(Exception triggered!)
    test_printf("Exception triggered!");
    vdl_Except(2) test_printf("Incorrect!");
    // expect(Handled!)
    vdl_Except(1) test_printf("Handled!");
    // expect(0)
    test_printf("%d", vdl_GetError());
    VDL_INTERNAL_GERROR = 1;
    // expect(Finally handled!)
    vdl_Finally() test_printf("Finally handled!");
    // expect(0)
    test_printf("%d", vdl_GetError());
    return 1;
}

_Noreturn void test_Abort(void)
{
    // echo
    echo("Test vdl_Abort");
    // exit(1)
    // expect(Program aborted!)
    test_printf_wrapper(vdl_Abort());
}
