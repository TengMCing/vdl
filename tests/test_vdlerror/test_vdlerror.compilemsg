In file included from test_vdlerror/test_vdlerror.c:7:
test_vdlerror/../test.h:31:17: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
        printf(f, ##__VA_ARGS__);                       \
                ^
test_vdlerror/../test.h:31:17: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
test_vdlerror/../test.h:31:17: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
test_vdlerror/../test.h:31:17: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
test_vdlerror/../test.h:31:17: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
test_vdlerror/../test.h:31:19: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
        printf(f, ##__VA_ARGS__);                       \
                  ^
test_vdlerror/../test.h:31:17: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
        printf(f, ##__VA_ARGS__);                       \
                ^
test_vdlerror/../test.h:31:19: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
        printf(f, ##__VA_ARGS__);                       \
                  ^
8 warnings generated.
