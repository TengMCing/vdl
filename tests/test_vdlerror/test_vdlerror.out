			Test vdl_GetError:
				Subtest 1: True
			Test vdl_internal_CheckError, vdl_Except and vdl_Finally:
				Subtest 1: Exception triggered!
				Subtest 2: Handled!
				Subtest 3: 0
				Subtest 4: Finally handled!
				Subtest 5: 0
			Test vdl_Abort
				Subtest 1: Program aborted!
